File: python/sportsdata_models/metadata.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from .catalogs import MODELS_ROOT, catalog_json_value_options, load_json


def metadata_path(sport: str, field: str) -> Path:
    return MODELS_ROOT / "metadata" / sport / f"{field}.json"


def load_metadata_field(sport: str, field: str) -> dict[str, Any]:
    return load_json(metadata_path(sport, field))


def metadata_json_value_options(sport: str, field: str) -> list[dict[str, str]]:
    metadata = load_metadata_field(sport, field)
    catalog = metadata["valueCatalog"]
    return catalog_json_value_options(
        catalog["scope"],
        catalog["name"],
        json_value_field=metadata.get("jsonValueField", "id"),
    )

File: web/src/metadata.ts
export interface MetadataField {
  id: string;
  sport: string;
  field: string;
  jsonKey: string;
  label: string;
  valueCatalog: {
    scope: string;
    name: string;
    path: string;
  };
  overrideCatalog?: string;
  manualValuesAllowed?: boolean;
  jsonValueField?: string;
  ui?: {
    jsonValueLabel?: string;
    jsonValueControl?: "select" | string;
  };
}

export const metadataPaths = {
  swimming: {
    distance: "models/metadata/swimming/distance.json",
  },
} as const;

export async function fetchMetadataField(path: string, baseUrl = ""): Promise<MetadataField> {
  const response = await fetch(`${baseUrl}${path}`);
  if (!response.ok) {
    throw new Error(`Unable to load metadata field ${path}: ${response.status}`);
  }
  return response.json() as Promise<MetadataField>;
}
